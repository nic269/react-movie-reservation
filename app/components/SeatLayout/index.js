import React, { memo } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import _upperCase from 'lodash/upperCase';
import _isArray from 'lodash/isArray';
import _get from 'lodash/get';

// Utils
import numberToVNDString from 'utils/number-to-vnd-string/number-to-vnd-string';
import { SEAT_TYPE } from 'utils/constants';

// Constants
import styles from './index.scss';

function SeatLayout(props) {
  const onSelect = seat => () => props.onSelect(seat);
  const { seatLayout, selectedIds, seatPrice } = props;
  return (
    <>
      <div className={styles.seatWrapper}>
        {_isArray(seatLayout) &&
          seatLayout.map(row => {
            const seats = _get(row, 'seats', []);
            const rowName = _get(row, 'name', '');
            const rowId = _get(row, 'id', '');

            return (
              <div className={styles.row} key={rowId}>
                <span className={styles.rowName}>{_upperCase(rowName)}</span>
                {seats.map((seat, idx) => {
                  const seatId = _get(seat, 'id', '');
                  const seatType = _get(seat, 'type', '');
                  const isDisabled =
                    seatType === SEAT_TYPE.empty ||
                    seatType === SEAT_TYPE.reserved;
                  const isSelected = selectedIds.includes(seatId);
                  const seatNo = idx + 1;

                  return (
                    <div
                      className={cx(styles.seat, {
                        [styles[seatType]]: true,
                        [styles.selected]: isSelected,
                      })}
                      disabled={isDisabled}
                      key={seatId}
                      onClick={isDisabled ? undefined : onSelect(seat)}
                      role="presentation"
                    >
                      {seatNo}
                    </div>
                  );
                })}
              </div>
            );
          })}
      </div>

      <div className={styles.instructions}>
        <div className={styles.instructionsStatus}>
          <div className={styles.instructionsRow}>
            <div
              className={cx(styles.seat, styles.rvd, styles.instructionsIcon)}
            />
            <span className={styles.instructionsText}>Đã đặt</span>
          </div>
          <div className={styles.instructionsRow}>
            <div
              className={cx(
                styles.seat,
                styles.selected,
                styles.instructionsIcon,
              )}
            />
            <span className={styles.instructionsText}>Đang chọn</span>
          </div>
        </div>
        <div className={styles.instructionsType}>
          {seatPrice[SEAT_TYPE.standard] && (
            <div className={styles.instructionsRow}>
              <div
                className={cx(styles.seat, styles.std, styles.instructionsIcon)}
              />
              <span className={styles.instructionsText}>
                Standard - {numberToVNDString(seatPrice[SEAT_TYPE.standard])}
              </span>
            </div>
          )}
          {seatPrice[SEAT_TYPE.vip] && (
            <div className={styles.instructionsRow}>
              <div
                className={cx(styles.seat, styles.vip, styles.instructionsIcon)}
              />
              <span className={styles.instructionsText}>
                VIP - {numberToVNDString(seatPrice[SEAT_TYPE.vip])}
              </span>
            </div>
          )}
          {seatPrice[SEAT_TYPE.deluxe] && (
            <div className={styles.instructionsRow}>
              <div
                className={cx(styles.seat, styles.dlx, styles.instructionsIcon)}
              />
              <span className={styles.instructionsText}>
                Deluxe - {numberToVNDString(seatPrice[SEAT_TYPE.deluxe])}
              </span>
            </div>
          )}
        </div>
      </div>
    </>
  );
}

SeatLayout.propTypes = {
  seatLayout: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      name: PropTypes.string,
      seats: PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.string,
          type: PropTypes.string,
        }),
      ),
    }),
  ),
  selectedIds: PropTypes.arrayOf(PropTypes.string),
  seatPrice: PropTypes.shape({
    [SEAT_TYPE.standard]: PropTypes.number,
    [SEAT_TYPE.vip]: PropTypes.number,
    [SEAT_TYPE.deluxe]: PropTypes.number,
  }),
  onSelect: PropTypes.func,
};

SeatLayout.defaultProps = {
  seatLayout: [],
  selectedIds: [],
  seatPrice: {},
  onSelect: () => {},
};

export default memo(SeatLayout);
