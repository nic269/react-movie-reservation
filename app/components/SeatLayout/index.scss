@import 'app/styles/theme';

.seatWrapper {
  display: grid;
  grid-row-gap: 5px;
}

.row {
  display: grid;
  grid-column-gap: 5px;
  grid-template-columns: repeat(15, 1fr);
}

.rowName {
  align-items: center;
  color: $standard-text-color;
  display: flex;
  justify-content: center;
}

.seat {
  align-items: center;
  border: 1px solid $standard-border-color;
  border-radius: 10px;
  color: transparent;
  cursor: pointer;
  display: flex;
  height: $seat-height;
  justify-content: center;

  &.std {
    border-color: $standard-border-color;
  }

  &.emp {
    visibility: hidden;
  }

  &.vip {
    border-color: $vip-border-color;
  }

  &.dlx {
    border-color: $deluxe-border-color;
  }

  &.selected {
    background-color: $selected-bg-color;
    color: $selected-text-color;
  }

  &.rvd {
    background-color: $reserved-bg-color;
    border-color: $reserved-border-color;
    cursor: default;
  }
}

.instructions {
  display: grid;
  grid-column-gap: 20px;
  grid-template-columns: 30% auto;
  margin-top: 40px;
  padding-left: 20px;
}

.instructionsStatus {
  display: grid;
  grid-row-gap: 10px;
  grid-template-rows: 20px 20px;
}

.instructionsType {
  display: grid;
  grid-row-gap: 10px;
  grid-template-rows: 20px 20px 20px;
}

.instructionsRow {
  align-items: center;
  display: flex;
}

.instructionsIcon {
  border-radius: 50%;
  display: inline-block;
  height: 20px;
  margin-right: 10px;
  width: 20px;
}

.instructionsText {
  color: $standard-text-color;
}
