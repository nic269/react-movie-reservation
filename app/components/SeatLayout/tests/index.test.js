/**
 *
 * Tests for SeatLayout
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render, fireEvent } from 'react-testing-library';
import 'jest-dom/extend-expect';

import SeatLayout from '../index';

describe('<SeatLayout />', () => {
  let props;

  beforeAll(() => {
    props = {
      seatLayout: [
        {
          id: 'row-a',
          seats: [
            {
              id: 'a1',
              type: 'std',
            },
            {
              id: 'a2',
              type: 'emp',
            },
          ],
        },
        {
          id: 'row-b',
          seats: [
            {
              id: 'b1',
              type: 'dlx',
            },
            {
              id: 'b2',
              type: 'rvd',
            },
          ],
        },
        {
          id: 'row-c',
          seats: [
            {
              id: 'c1',
              type: 'vip',
            },
            {
              id: 'c2',
              type: 'vip',
            },
          ],
        },
      ],
      selectedIds: ['a1'],
      seatPrice: {
        std: 10000,
        dlx: 20000,
        vip: 15000,
      },
      onSelect: jest.fn(),
    };
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(<SeatLayout {...props} />);
    expect(firstChild).toMatchSnapshot();
  });

  it('onSelect', () => {
    const { getByTestId } = render(<SeatLayout {...props} />);

    fireEvent.click(getByTestId('seat-layout-a1'));
    expect(props.onSelect).toHaveBeenCalled();
  });
});
