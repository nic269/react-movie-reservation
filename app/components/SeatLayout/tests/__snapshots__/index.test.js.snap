// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SeatLayout /> Should render and match the snapshot 1`] = `
<div>
  <div>
    <span />
    <div
      class="undefined"
      data-testid="seat-layout-a1"
      role="presentation"
    >
      1
    </div>
    <div
      class=""
      data-testid="seat-layout-a2"
      disabled=""
      role="presentation"
    >
      2
    </div>
  </div>
  <div>
    <span />
    <div
      class=""
      data-testid="seat-layout-b1"
      role="presentation"
    >
      1
    </div>
    <div
      class=""
      data-testid="seat-layout-b2"
      disabled=""
      role="presentation"
    >
      2
    </div>
  </div>
  <div>
    <span />
    <div
      class=""
      data-testid="seat-layout-c1"
      role="presentation"
    >
      1
    </div>
    <div
      class=""
      data-testid="seat-layout-c2"
      role="presentation"
    >
      2
    </div>
  </div>
</div>
`;
